### PR TITLE
Introduce the ability to generate AppCDS

### PR DESCRIPTION
**Pull request by geoand**

### CODE REVIEW

**Comment by geoand** [comment]
@stuartwdouglas mind taking a quick look at this and let me know if you agree with the approach?

If so, I'll proceed to add tests for all three 3 jars types (for tests the JVM needs to be launched with `-Xshare:on` which will ensure the JVM will fail if the AppCDS were not usable) and then add this integration into Jib as well.

**Comment by geoand** [comment]
@loicmathieu mind giving this a spin?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    /**
+     * @return The path of the created app-cds.jsa file or null of the file was not created
+     */
+    private Path createAppCDS(JarBuildItem jarResult, String javaBin,
+            Path classesLstPath, boolean isFastFar) {
+
+        Path workingDirectory = isFastFar ? jarResult.getPath().getParent().getParent() : jarResult.getPath().getParent();
+        Path appCDSPath = workingDirectory.resolve("app-cds.jsa");
```
Comment (stuartwdouglas) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:148]: For fast-jar this should go into the actual fast-jar package.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+        command.add("--class-path");
+        command.add(jarResult.getPath().getFileName().toString());
+
+        if (log.isDebugEnabled()) {
+            log.debugf("Launching command: '%s' to create final AppCDS", String.join(" ", command));
+        }
+
+        int exitCode;
+        try {
+            exitCode = new ProcessBuilder(command)
```
Comment (stuartwdouglas) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:168]: I think this approach is going to fail in a lot of cases, as you need all the production services to be present.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
I like the concept, but in practice I think there will be the potential for a lot of problems, including:

- Big potential for mismatching versions between container and local host. I am not sure what happens in this case? Do you end up with something that won't start?
- Production databases need to be available when building, as the build runs in prod profile. There is a good chance in most cases this just won't be possible.

I think that we will likely want an approach that just loads all the classes, rather than explicitly starting the application, as even though it generates a larger image it avoids the services problem.

I also think we will want to generate this as part of image creation, so you know that you will always have the correct JVM version. I think this is easy enough with a Dockerfile, not sure how hard it will be for jib.

**Comment by geoand** [comment]
> I like the concept, but in practice I think there will be the potential for a lot of problems, including:
> 
> * Big potential for mismatching versions between container and local host. I am not sure what happens in this case? Do you end up with something that won't start?

You are right that you have to have exact same version of Java for the AppCDS to work. However, the default behavior of the JVM is to just ignore the AppCDS if there is an issue for any reason. For the JVM to fail when there is an issue with AppCDS, you need to have `-Xshare:on` set.

> * Production databases need to be available when building, as the build runs in prod profile. There is a good chance in most cases this just won't be possible.
> 
> I think that we will likely want an approach that just loads all the classes, rather than explicitly starting the application, as even though it generates a larger image it avoids the services problem.
> 

Yeah, that's definitely the safer approach and for the dependencies it's easy. But what do we do about the JDK classes?

> I also think we will want to generate this as part of image creation, so you know that you will always have the correct JVM version. I think this is easy enough with a Dockerfile, not sure how hard it will be for Jib.

Here is the thing about this approach: You **need** to have docker installed to make this work, in order to run the JVM of the built container image to generate the AppCDS and copy it out of the container.
That doesn't change with Jib, because with Jib all we can do is build the container image, we still need docker to actually run it so we can generate the AppCDS.

In conclusion I agree with you that the idea is interesting, but we do need to think about how to provide the best experience. 

cc @maxandersen

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java (hunk 1 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:148]: The way it's done, it ends up in the directory that `quarkus-run.jar` exists. Is that what you meant, or something else?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java (hunk 2 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:168]: It could certainly fail. I'm not saying it's a good solution - because it isn't, but even like this you still get a list classes :)

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java (hunk 1 above)
Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:148]: That is good, I missread this bit.

Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:148]: That is good, I missread this bit.

**Comment by stuartwdouglas** [comment]
Jdk classes should be mostly fine, as loading all the classes will run all the static init code.

**Comment by geoand** [comment]
> Jdk classes should be mostly fine, as loading all the classes will run all the static init code.

I am certainly missing something, as I don't get you mean here :)

My question was how would we add the JDK classes to `classes.lst`. Do we just always add a bunch of classes we know will be used?

**Comment by stuartwdouglas** [comment]
What I am talking about is:
- Create a file at build time that lists all the classes in the runtime deps
- Create a new main() method that loads the file, uses Class.forName() on every class, then exits
- this Class.forName() will run the generated static init code, which will load most(all?) of the JDK classes we care about

**Comment by geoand** [comment]
Gotcha, thanks for the explanation

**Comment by geoand** [comment]
I have to prepare for a demo today and I'll most likely be off tomorrow, so I'll probably start looking at this again at the end of the week.

**Comment by loicmathieu** [comment]
Starting with Java 13 this can be done in one path using dynamic CDS archive (no need to create the classlist upfront): http://openjdk.java.net/jeps/350

I do love the fact that it didn't depends on docker on anything else, but as Stuart said it could fail in a lot of way. We will also need a way to integrate it in our generated Dockerfile.

I need to check again why I use the uberjar in my blogpost, I'm sure the normal jar didn't worked for jlink but I'm not sure for AppCDS. Maybe there is an issue with the normal jar and AppCDS not detecting all classes due to missing libs in the classloader used by AppCDS to create the class list ...

**Comment by geoand** [comment]
> Starting with Java 13 this can be done in one path using dynamic CDS archive (no need to create the classlist upfront): http://openjdk.java.net/jeps/350
> 

Yeah. I was thinking thas if we get this in, then as an enhancement later on we could use this if Java 13+ is being used.

> I do love the fact that it didn't depends on docker on anything else, but as Stuart said it could fail in a lot of way. We will also need a way to integrate it in our generated Dockerfile.
> 

This is exactly why I did it this way - so users don't **need** the container image to take advantage of this - although there are a lot of caveats.

> I need to check again why I use the uberjar in my blogpost, I'm sure the normal jar didn't worked for jlink but I'm not sure for AppCDS. Maybe there is an issue with the normal jar and AppCDS not detecting all classes due to missing libs in the classloader used by AppCDS to create the class list ...

I was able to make it make it work for all types of jars - although for `fast-jar` the speedup improvement was very small. For `uber-jar` and `legacy-jar` the speedup was very good.

**Comment by loicmathieu** [comment]
> although for fast-jar the speedup improvement was very small

This is what I fear, it means the classlist didn't contains all the needed classes. You can enable class loading logging to check this: `-Xlog:class+load:file=cds.log`.
Classes lodading from AppCDS will be logged like this:
```
[info][class,load] java.lang.Object source: shared objects file
```

**Comment by geoand** [comment]
> > although for fast-jar the speedup improvement was very small
> 
> This is what I fear, it means the classlist didn't contains all the needed classes. You can enable class loading logging to check this: `-Xlog:class+load:file=cds.log`.
> Classes lodading from AppCDS will be logged like this:
> 
> ```
> [info][class,load] java.lang.Object source: shared objects file
> ```

Yeah, I was doing that when I was trying to get all types of jars to work. I just didn't have time to explore why `fast-jar` in particular did not see a proper speedup (even though the JVM was applying AppCDS). My guess is that it has something to do with the custom ClassLoader that `fast-jar` uses, but I could be completely off.

**Comment by geoand** [comment]
I just got this crazy idea (although it's not definitely not very user friendly):

What if we used the report of the used classes that GraalVM outputs as the classes list? There is no way we are going to get a more accurate list of classes than that. If we did that, all we would have to do is feed that list into `java -Xshare:dump ...` and out comes the AppCDS file.

I admit that from a usability standpoint it's not great, but I wanted to put it out there to see what y'all think.

P.S. It would be nice if we didn't have to actually run the whole native-image build process, but instead only utilize it to do static analysis and spit out the used classes report.

**Comment by loicmathieu** [comment]
@geoand `native-image` is definitly too slow and would include a lot of classes not really used by the application as it follow all possible code flow. Moreover, we are only interested on the class used during the bootstraping, classes used during the run of the application and not discovered by AppCDS will be loaded over time and it's OK as it will balance the AppCDS size and the bootstrap time.

**Comment by geoand** [comment]
@loicmathieu actually if native image analysis wasn't so slow and require extra tooling, I think it would be the perfect, because it discovers all the classes loaded in the static init phase - I don't see how you can get a better set of classes actually used at runtime without starting the application itself.

But I agree that the current slowness of the process is prohibitive... Which is why I wonder if there is a way for us to only run the static init phase up until the used classes report is generated

**Comment by geoand** [comment]
I'll implement @stuartwdouglas's proposal hopefully next week when done with PTO and I've taken care of some other higher priority loose ends.

**Comment by geoand** [comment]
I overhauled the PR. Now we don't need to start the application at all and the AppCDS are created by generating a `classes.lst` file from the JDK's `lib/classlist` and all the classes of the application and its dependencies.

This process doesn't provide any means of adding the generated AppCDS into a generated container as that requires building a container, running it (to make use of the exact Java version) to create the AppCDS and then generating another container once we have the AppCDS.
It is doable (with a Docker builder image it should be pretty easy), but I would prefer we first have the raw capability in and get some feedback.

WDYT @stuartwdouglas? If you agree, I'll add some tests and documentation before taking the PR out of draft.

**Comment by gunnarmorling** [comment]
> I overhauled the PR. Now we don't need to start the application at all and the classes list is created by generating a classes.lst file from the combination of JDK's lib/classlist and looking at all the classes of the application and its dependencies.

Sounds like a good approach to start with. We'd loose the ability to have different "profiles" for different kinds of workloads, but that seems acceptable. Also you'd probably miss things like Lambda proxy classes (as CDS-able as of JDK 15-ea-b27), but that seems acceptable, too.

**Comment by geoand** [comment]
> > I overhauled the PR. Now we don't need to start the application at all and the classes list is created by generating a classes.lst file from the combination of JDK's lib/classlist and looking at all the classes of the application and its dependencies.
> 
> Sounds like a good approach to start with. We'd loose the ability to have different "profiles" for different kinds of workloads, but that seems acceptable. Also you'd probably miss things like Lambda proxy classes (as CDS-able as of JDK 15-ea-b27), but that seems acceptable, too.

Yeah, it can and should definitely be improved in follow up iterations

**Comment by loicmathieu** [comment]
With this approach you will create a real big classlist as you include all the classes from all the libraries ! You also inlude all the JDK classes.
I'm very concerned about this as it will take time to create the CDS archive and could have a huge impact on the CDS size, the RSS memory needed to load it and the startup time needed to load it.

Did you try it ? Do you know it's impact on disk size, RSS and startup time ?

**Comment by geoand** [comment]
> With this approach you will create a real big classlist as you include all the classes from all the libraries ! You also inlude all the JDK classes.

Yup, we definitely want the JDK classes. Short of running the application (which we don't want to do), there is no other way of getting the list of classes.

> I'm very concerned about this as it will take time to create the CDS archive and could have a huge impact on the CDS size, the RSS memory needed to load it and the startup time needed to load it.
> 
> Did you try it ? Do you know it's impact on disk size, RSS and startup time ?

I did try and and it worked well for what I tested. I would invite you to test this on the application you have in mind and see for yourself :)

**Comment by loicmathieu** [comment]
> I did try and and it worked well for what I tested. I would invite you to test this on the application you have in mind and see for yourself :)

The question is not if it works but the overhead of disk size and memory consumption reagrding an implementation that build the classlist based on the application run.

I'll try to test it Friday, I should have some spare time ...

**Comment by geoand** [comment]
> > I did try and and it worked well for what I tested. I would invite you to test this on the application you have in mind and see for yourself :)
> 
> The question is not if it works but the overhead of disk size and memory consumption reagrding an implementation that build the classlist based on the application run.
> 

That's what I meant...

**Comment by geoand** [comment]
> > I did try and and it worked well for what I tested. I would invite you to test this on the application you have in mind and see for yourself :)
> 
> The question is not if it works but the overhead of disk size and memory consumption reagrding an implementation that build the classlist based on the application run.
> 
> I'll try to test it Friday, I should have some spare time ...

So I did try this out on a larger example, and indeed the RSS does go up a lot... 
The startup time improvement is great (something like 60% of the initial start up time) but the RSS increase seems very high - something like 25%. 
Disk size of the file is also very large, like 150MB :(.

In light of this I am wondering if this approach does make sense after all...

**Comment by gunnarmorling** [comment]
RSS might be a false alert, see
https://twitter.com/gunnarmorling/status/1271745125920759808.

But the disk size usage is prohibitive indeed.

**Comment by geoand** [comment]
> RSS might be a false alert, see https://twitter.com/gunnarmorling/status/1271745125920759808. 

Excellent info, thanks! 
I did run a manual GC and memory sure came down. But the same applies when I ran a manual GC on the application when it was not using AppCDS.
The only way the RSS was comparable was when the application with AppCDS was GCed and the application without AppCDS was not GCed.
So I am really not sure what to make of it...

**Comment by loicmathieu** [comment]
> RSS might be a false alert, see

Yes and no ... the JVM will use this much memory to load the CDS Archive, so in a constrained environment we will need a larger container memory size.
But after a full GC, it will go back to normal.
Can you try again and trigger a full GC (via JMX) to check this assumption ?

**Comment by geoand** [comment]
I wonder if it would make sense for the purpose of creating AppCDS to generate a main that only runs static-init and nothing else...

**Comment by loicmathieu** [comment]
I was playin with this lately ;)

If you create a main that return instantly you still load the apps, opens the HTTP port, then exits instantly.

Ideally, you should trigger an HTTP call to the main endpoint (at least) to load more classes but it should works.

I needed it for something else and facing some issue (NPE) but yeah, having an empty main will load most of the Quarkus bootstraping code and this is what we want to include inside the CDS archive.
If the endpoint is not hit, you will still have some part of the bootstraping code not listed but this is a good start.

**Comment by geoand** [comment]
> I was playin with this lately ;)
> 
> If you create a main that return instantly you still load the apps, opens the HTTP port, then exits instantly.
> 
> Ideally, you should trigger an HTTP call to the main endpoint (at least) to load more classes but it should works.
> 
> I needed it for something else and facing some issue (NPE) but yeah, having an empty main will load most of the Quarkus bootstraping code and this is what we want to include inside the CDS archive.
> If the endpoint is not hit, you will still have some part of the bootstraping code not listed but this is a good start.

Yeah, that's what I'm thinking. Since loading all classes provides to heavy an output and we really don't want to launch the regular application, I think doing an only static-init main would make sense.
If @stuartwdouglas agrees, I'll go ahead and do that.

**Comment by geoand** [comment]
Found some time to do a 3rd implementation.

So basically what happens now is that we do launch the application to get the list of classes, but the the whole application. Only classes loaded from static init and Quarkus generated or transformed classes are taken into account.

The startup gains aren't as good as the previous implementation, but the footprint of both the generated AppCDS file and the RSS is dramatically less.
I think it's a good compromise.

**Comment by gunnarmorling** [comment]
> Yes and no ... the JVM will use this much memory to load the CDS Archive, so in a constrained environment we will need a larger container memory size.

I don't think that's true necessarily. You can give it less heap
space, it will then just GC early on. I could run my example app with
64MB heap without problem. Going down to 32MB caused quite some GC
overhead then.

**Comment by loicmathieu** [comment]
> You can give it less heap space, it will then just GC early on.

Yes, but RSS is not ony heap size, I don't know how the CDS archive is read but it can uses native memory to read it so it will impact RSS size and not a lot of heap size ...

**Comment by geoand** [comment]
@stuartwdouglas since we are getting close to the cut for 1.6, I would be really grateful if you could weigh in on whether the latest implementation can be added (as experimental).

Thanks

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/PackageConfig.java [hunk 3]
```diff
@@ -79,4 +79,30 @@
...
+    public boolean isAnyJarType() {
+        return (type.equalsIgnoreCase(PackageConfig.LEGACY) ||
+                type.equalsIgnoreCase(PackageConfig.JAR) ||
+                type.equalsIgnoreCase(PackageConfig.FAST_JAR) ||
+                type.equalsIgnoreCase(PackageConfig.UBER_JAR)) ||
+                type.equalsIgnoreCase(PackageConfig.MUTABLE_JAR);
+    }
+
+    public boolean isFastJar() {
+        return type.equalsIgnoreCase(PackageConfig.FAST_JAR);
```
Comment (stuartwdouglas) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/PackageConfig.java:101]: MUTABLE_JAR is also a type of FAST_JAR.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        log.debug("'" + CLASSES_LIST_FILE_NAME + "' successfully created.");
+
+        log.info("Launching AppCDS creation process.");
+        Path appCDSPath = createAppCDS(jarResult, javaBinStr, classesLstPath, packageConfig.isFastJar());
+        if (appCDSPath == null) {
+            log.warn("Unable to create AppCDS.");
+            return;
+        }
+
+        log.infof(
```
Comment (stuartwdouglas) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:88]: Maybe also include info about the JDK used to build it in this message?

**Review by stuartwdouglas — APPROVED** [review]
LGTM

**Comment by geoand** [comment]
PR updated per review comments. I also added a test.

I still need to write some docs, but I'll do that early next week.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/PackageConfig.java [hunk 5]
```diff
@@ -79,4 +79,31 @@
      */
     @ConfigItem
     public Optional<String> outputName;
+
+    /**
+     * Whether to automate the creation of AppCDS. This has not effect when a native binary is needed and will be ignored in
+     * that case.
+     * Furthermore, this options only works for Java 11+ and is considered experimental for the time being.
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/PackageConfig.java:86]:
```suggestion
     * Furthermore, this option only works for Java 11+ and is considered experimental for the time being.
```

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/builditem/AppCDSRequestedBuildItem.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package io.quarkus.deployment.pkg.builditem;
+
+import java.nio.file.Path;
+
+import io.quarkus.builder.item.SimpleBuildItem;
+
+/**
+ * Build item to notify the indicate to the the various steps that AppCDS generation
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/builditem/AppCDSRequestedBuildItem.java:8]:
```suggestion
 * Build item to indicate to the various steps that AppCDS generation
```

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+            return;
+        }
+
+        Path classesLstPath = createClassesLst(packageConfig, jarResult, javaBinStr, appCDSDir);
+        if (classesLstPath == null) {
+            log.warn("Unable to create AppCDS because " + CLASSES_LIST_FILE_NAME + " was not created.");
+            return;
+        }
+
+        log.debug("'" + CLASSES_LIST_FILE_NAME + "' successfully created.");
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:80]: debugf maybe?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+                        "run the application jar from its directory and also add the '-XX:SharedArchiveFile=app-cds.jsa' " +
+                        "JVM flag.\nMoreover, make sure to use the exact same Java version (%s) to run the application as was used to build it.",
+                appCDSPath.toAbsolutePath().toString(), System.getProperty("java.version"));
+
+        appCDS.produce(new AppCDSResultBuildItem(appCDSPath));
+        artifactResult.produce(new ArtifactResultBuildItem(appCDSPath, "appCDS", Collections.emptyMap()));
+    }
+
+    /**
+     * @return The path of the created classes.lst file or null of the file was not created
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:100]:
```suggestion
     * @return The path of the created classes.lst file or null if the file was not created
```

**Review by gsmet — DISMISSED** [review]
Did some proofreading.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+        if (exitCode != 0) {
+            log.debugf("The process that was supposed to create AppCDS exited with error code: %d.", exitCode);
+            return null;
+        }
+
+        return appCDSDir.resolve(CLASSES_LIST_FILE_NAME);
+    }
+
+    /**
+     * @return The path of the created app-cds.jsa file or null of the file was not created
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:146]:
```suggestion
     * @return The path of the created app-cds.jsa file or null if the file was not created
```

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java (hunk 7 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java:80]: Good idea

**Comment by gsmet** [comment]
Could you write a paragraph for the release notes about what it brings to the plate and how to use it? Just paste it in a comment. Thanks!

**Comment by geoand** [comment]
```
Quarkus 1.6 brings experimental support for automatically a producing [Application Class-Data Sharing (AppCDS)](http://openjdk.java.net/jeps/310) file. AppCDS essentially allow an application to start faster by reducing the work the JVM needs to do at startup time and Quarkus can now automatically create the AppCDS file as part of the build by simply adding `quarkus.package.create-appcds=true`.
There are various limitations the JVM imposes on the use of AppCDS, so make sure to read our guide before using this feature.
```
